extendeddatacrossword.go

```diff
@@ -101,177 +101,208 @@ func RepairExtendedDataSquare(
-							vectorData = eds.Row(uint(i))
-						case column:
-							r = j
-							c = i
-							vectorData = eds.Column(uint(i))
-						default:
-							panic(fmt.Sprintf("invalid mode %d", mode))
-						}
-						if bitMask.Get(r, c) {
-							// As guaranteed by the bitMask, vectorData can't be nil here:
-							shares[j] = vectorData[j]
-						}
-					}
-
-					// Attempt rebuild
-					rebuiltShares, err := codec.Decode(shares)
-					if err != nil {
-						// repair unsuccessful
-						solved = false
-						continue
-					}
-
-					progressMade = true
-
-					if isExtendedPartIncomplete {
-						// If needed, rebuild the parity shares too.
-						rebuiltExtendedShares, err := codec.Encode(rebuiltShares[0:eds.originalDataWidth])
-						if err != nil {
-							return err
-						}
-						startIndex := len(rebuiltExtendedShares) - int(eds.originalDataWidth)
-						rebuiltShares = append(
-							rebuiltShares[0:eds.originalDataWidth],
-							rebuiltExtendedShares[startIndex:]...,
-						)
-					} else {
-						// Otherwise copy them from the EDS.
-						startIndex := len(shares) - int(eds.originalDataWidth)
-						rebuiltShares = append(
-							rebuiltShares[0:eds.originalDataWidth],
-							shares[startIndex:]...,
-						)
-					}
-
-					// Check that rebuilt shares matches appropriate root
-					err = eds.verifyAgainstRoots(rowRoots, columnRoots, mode, uint(i), bitMask, rebuiltShares)
-					if err != nil {
-						return err
-					}
-
-					// Check that newly completed orthogonal vectors match their new merkle roots
-					for j := 0; j < int(eds.width); j++ {
-						switch mode {
-						case row:
-							if !bitMask.Get(i, j) &&
-								bitMask.ColumnIsOne(j) {
-								err := eds.verifyAgainstRoots(rowRoots, columnRoots, column, uint(j), bitMask, rebuiltShares)
-								if err != nil {
-									return err
-								}
-							}
-
-						case column:
-							if !bitMask.Get(j, i) &&
-								bitMask.RowIsOne(j) {
-								err := eds.verifyAgainstRoots(rowRoots, columnRoots, row, uint(j), bitMask, rebuiltShares)
-								if err != nil {
-									return err
-								}
-							}
-
-						default:
-							panic(fmt.Sprintf("invalid mode %d", mode))
-						}
-					}
-
-					// Set vector mask to true
-					switch mode {
-					case row:
-						for j := 0; j < int(eds.width); j++ {
-							bitMask.Set(i, j)
-						}
-					case column:
-						for j := 0; j < int(eds.width); j++ {
-							bitMask.Set(j, i)
-						}
-					default:
-						panic(fmt.Sprintf("invalid mode %d", mode))
-					}
-
-					// Insert rebuilt shares into square.
-					for p, s := range rebuiltShares {
-						switch mode {
-						case row:
-							eds.setCell(uint(i), uint(p), s)
-						case column:
-							eds.setCell(uint(p), uint(i), s)
-						default:
-							panic(fmt.Sprintf("invalid mode %d", mode))
-						}
-					}
-				}
+			solvedRow, progressMadeRow, err := eds.solveCrosswordRow(i, rowRoots, columnRoots, bitMask, codec)
+			if err != nil {
+				return err
+			}
```

Comment:
We should open an issue about covering all the non-happy paths if it doesn't exist already.